File: conformance/tests/schnorr-zkp/schnorr-zkp.runar.zig
const runar = @import("runar");

pub const SchnorrZKP = struct {
    pub const Contract = runar.SmartContract;

    pubKey: runar.Point,

    pub fn init(pubKey: runar.Point) SchnorrZKP {
        return .{ .pubKey = pubKey };
    }

    pub fn verify(self: *const SchnorrZKP, rPoint: runar.Point, s: i64) void {
        runar.assert(runar.ecOnCurve(rPoint));

        const e = runar.bin2num(runar.hash256(runar.cat(rPoint, self.pubKey)));
        const sG = runar.ecMulGen(s);
        const eP = runar.ecMul(self.pubKey, e);
        const rhs = runar.ecAdd(rPoint, eP);

        runar.assert(runar.ecPointX(sG) == runar.ecPointX(rhs));
        runar.assert(runar.ecPointY(sG) == runar.ecPointY(rhs));
    }
};

File: conformance/tests/schnorr-zkp/schnorr-zkp.runar.go
//go:build ignore

package contract

import "runar"

type SchnorrZKP struct {
	runar.SmartContract
	PubKey runar.Point `runar:"readonly"`
}

func (c *SchnorrZKP) Verify(rPoint runar.Point, s runar.Bigint) {
	runar.Assert(runar.EcOnCurve(rPoint))
	e := runar.Bin2Num(runar.Hash256(runar.Cat(rPoint, c.PubKey)))
	sG := runar.EcMulGen(s)
	eP := runar.EcMul(c.PubKey, e)
	rhs := runar.EcAdd(rPoint, eP)
	runar.Assert(runar.EcPointX(sG) == runar.EcPointX(rhs))
	runar.Assert(runar.EcPointY(sG) == runar.EcPointY(rhs))
}
